### ISSUE TITLE
Makumba should provide a filter to delete spaces between tags

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#520]
**Reported by @rudolfmayer on 29 Jul 2007 17:30 UTC**
There are several pages in private area that do not return a HTML page, but binary files, e.g. showing a picture, or exporting data to CSV or excel format.

the problem in those pages is that they must no contain a single space, as that would trigger JSP to write to the output stream. that requires however to write all tags next to each other, or make line.breaks only in between scriplets, e.g.

<some tag><% 
%><some other tag>

one example is e.g. http://private.best.eu.org/events/applicantListExport.jspx

code becomes very much unreadable like this, and it is difficult for beginner programmers to understand why it has to be like this, and how to debug it.

therefore, even though not a directly makumba related problem, some support for this would make sense, one solution could be to implement a filter that would delete all whitespace between tags.
one question is how this filter should be activated (enabling it by default is imho not a good solution).

[such pages as servlets or at least via functions in BL would be an option, and the preferred solution actually, but that is again not easy for the average programmer. also, doing nested mak:lists is not optimised in BL, afaik, so some advantages of the makumba taglib get lost](implementing)

Migrated-From: http://trac.makumba.org/ticket/972

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 29 Apr 2010 22:35 UTC**
this is very useful also when u do HTML because sometimes even there you don't want the space.

one solution is the global

<mak:noWhitespace>

some content

</mak:noWhitespace>

which would remove all whitespace except for the one inside ' '. i would guess this should be available in some jsp taglib out there?

----

the other solution is the local

<mak:value ....> <mak:eatWhitespace /> 

</mak:list>

then all whitespace between the end of value and the doEndTag of /list would be removed.

**Comment by ghost** [comment]
**Comment by @filipkis on 1 May 2010 10:55 UTC**
(In reply to comment #1) 
> one solution is the global
> 
> <mak:noWhitespace>
> 
> some content
> 
> </mak:noWhitespace>
> 
> which would remove all whitespace except for the one inside ' '. i would guess
> this should be available in some jsp taglib out there?

I like this approach! 

However if there is a simple jsp taglib that supports that, I'm not sure makumba should have a special tag for it. Unless it brings some real benefit (reducing the code, some linking to makumba values that should otherwise be done manually etc.).


> ----
> 
> the other solution is the local
> 
> <mak:value ....> <mak:eatWhitespace />
> 
> </mak:list>
> 
> then all whitespace between the end of value and the doEndTag of /list would be
> removed.

I don't really like this solution. It somehow looks weird. It does something after the end of it's tag and I would find that as normal behavior. I think that above nowWithespace would solve most of the cases anyhow.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 2 May 2010 18:18 UTC**
(In reply to comment #2)
> However if there is a simple jsp taglib that supports that, I'm not sure
> makumba should have a special tag for it. Unless it brings some real benefit
> (reducing the code, some linking to makumba values that should otherwise be
> done manually etc.).

usually when i say that there might be a taglib out there, i as well mean that we should not do the same thing. i also mean that those who want to help could do a search "out there" :)

it turns out that JSP itself does have this feature, or a version of it
http://raibledesigns.com/rd/entry/trim_spaces_in_your_jsp

also there is a "trim filter" that can do this.

> I don't really like this solution. It somehow looks weird. It does something
> after the end of it's tag and I would find that as normal behavior. I think
> that above nowWithespace would solve most of the cases anyhow.

well not really because as you see in the link above the complexity of the whitespace issue is just funny.

in any case i also think that the local solution is weak because you simply have to write too much to just achieve whitespace removal.

now about the complexities: my basic assumption is that whitespace is added for code clarity reasons:
- spaces before the tag, for alignment. the tag is then usually the only one on that line
- the newline after the tag
- empty lines
- if two tags are  on the same line, with space between them, i would not remove that.

the question is of course whether this all is done at HTML or JSP taglib level.
if any of the existing solutions eliminates these (and not much more) then we could gladly use it.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 2 May 2010 19:57 UTC**
even more fun
http://mindprod.com/project/htmlcompactor.html
does quite good analysis

another issue is whether this should be usable for just page fragments, for the whole page, or for the whole site.

it seems that there are several classes of such tools depending on whether
- they work at JSP tag level _or_ at HTML/text level
- they preserve the HTML rendering (i.e. what you see in the browser is the same though the HTML changes), _or_ assume that some output comes from code clarity issues, and therefore it does make sense to remove what otherwise would be a HTML space, thus changing the HTML rendering

i think it is the latter (address code formatting) that is the object of this bug.

i guess the HTML-only tools practically have to preserve the HTML output. so filters will not be able to fix this bug. the only tool i know that uses JSP knowledge is the JSP trimSpace but there are also complaints that it trims too much.

a JSP level tool that changes the HTML rendering to eliminate extra space that's due to code formatting would be used consciously by the JSP programmer, and as long as the JSP programmer understands how it works, it will not create undesired effects.

for example:

     <mak:value expr=....>
  </td>
</tr>

the tool that i think this bug requests should remove 
- the spaces from the beginning of the line to mak:value 
- the whitespace until the /td including the newline

in this case it makes a lot of sense to operate at jsp taglib level and not at HTML level because it does not matter what the mak:value returns!

one important issue here is that the <mak:value> and the /td are on different lines. indeed, if they were on the same line the programmer would have had all control without the code looking ugly.

------

so one way to do this may be to make a clever stream (a decorator plugged in by a filter) that removes 
1) only whitespace that contains newlines and
2) has at start and/or end a makumba tag (either mak:xyz or /mak:xyz)

(2) can be relaxed for "any non-whitespace content" however that would need to check for the <pre> case, and probably other cases

i would turn this on from Makumba.conf (per site or per site portions) and not via a tag because it would make it much more convenient to use.

also a way to go around this should be provided. i guess using some special char like &emspace; would do? 

in other words

<mak:value/>
<mak:value/>
will be equivalent with <mak:value/><mak:value/>
to add a space one would need to use
<mak:value/> <mak:value/> 
or 
<mak:value/>&emspace;
<mak:value/>

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 2 May 2010 22:08 UTC**
i checked trimSpace at the JSP servlet in conf/web.xml. it requires work/ cleanup and tomcat restart.

        <init-param>
        <param-name>trimSpaces</param-name>
        <param-value>true</param-value>
    </init-param>


this is a tag example, nicely formatted with Eclipse. i don't do too many of those precisely due to whitespace issues, but i got this one from Rudi:

			<td><mak:criterion fields="parti" id="gigi">
				<mak:searchField forceInputStyle="multiple" size="8"
					onChange="this.form.submit();">
					<mak:option>Toate</mak:option>
					<mak:list from="dox.Parte dp" where="dp.nume not in ($mdt)">
						<mak:option value="dp">
							<mak:value expr="dp.nume" />
						</mak:option>
					</mak:list>
				</mak:searchField>
			</mak:criterion></td>

the result before trimSpace: 

                     <td> 
				<SELECT MULTIPLE NAME="parti" SIZE="8"  onChange="this.form.submit();" id="parti_form1" > 
	<OPTION VALUE="">Toate</OPTION> 
	<OPTION VALUE="24ojiz7"> 
							guru SC
						</OPTION> 
	<OPTION VALUE="43pnkya"> 
							alfabeta
						</OPTION> 
	<OPTION VALUE="24ojiz9"> 
							gamma
						</OPTION> 
	<OPTION VALUE="43pnkyc"> 
							zuzu
						</OPTION> 
</SELECT> 
			</td> 

the result after trimSpace:

                   <td><SELECT MULTIPLE NAME="parti" SIZE="8"  onChange="this.form.submit();" id="parti_form1" > 
	<OPTION VALUE="">Toate</OPTION> 
	<OPTION VALUE="24ojiz7">guru SC</OPTION> 
	<OPTION VALUE="43pnkya">alfabeta</OPTION> 
	<OPTION VALUE="24ojiz9">gamma</OPTION> 
	<OPTION VALUE="43pnkyc">zuzu</OPTION> 
</SELECT></td> 

however if i transform the middle mak.value in a "double" one

							<mak:value expr="dp.nume" /> <mak:value expr="dp.TS_create"/>

the result is like

	<OPTION VALUE="24ojiz7">guru SC2010-02-02 14:35:21.0</OPTION> 
	<OPTION VALUE="43pnkya">alfabeta2010-02-02 14:48:21.0</OPTION> 
	<OPTION VALUE="24ojiz9">gamma2010-02-02 14:48:58.0</OPTION> 
	<OPTION VALUE="43pnkyc">zuzu2010-02-02 23:17:01.0</OPTION> 

i.e. the space between the two mak:values is lost. using a newline instead (like eclipse would format them) has no effect. using &emsp; does the job, as well as inserting a comment like

							<mak:value expr="dp.nume" />
							
							<mak:value expr="dp.TS_create" />

also any separator between the two mak:values will lead to _all_ whitespace between them being preserved, while the whitespace between the mak:values and the surrounding mak:option is removed. 

------
an important player here is eclipse's Format Source. for the two successive mak:values, it will generate mak:value newline mak:value as above. in that case, trimSpaces and the algorithm i propose at comment 4 would have exactly the same effect!

in fact if you have two mak:values and Eclipse formats them on two lines, it is impossible to tell algorithmically whether you wanted a space between them or not, no matter what tool you use. if it's a mak:value after a mak:list or a /mak:list after a mak:value, it is much easier to assume that you don't want space. the same goes for td mak:value, mak:value /td.

so in most cases you don't want the whitespace between makumba (and other JSP) tags and other JSP or HTML tags, except between two mak:values where it's a tie. 

-----

_therefore_ my proposal is to make a tool that looks for successive mak:values and mak:inputs, or other successions of similar tags, fix these cases, and use trimSpaces from then on! 

note that the fixing can be automatic! 

- for binary cases as in comment 0, they don't have space in the output anyway, so this is no problem. the file can be nicely formatted with eclipse and the trimSpaces will do!

- for HTML 
 <mak:value > <mak:value> will turn into <mak:value ><%=" "%><mak:value>
back-to-back <mak:value ><mak:value> can be left as is

- for plain text, like HTML plus: any newline between a a JSP tag and another tag needs to be replaced with 
<%="\n" %>

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 3 May 2010 08:07 UTC**
the trimSpaces solution is not per-webapp, but can be used per page, because it is of course possible to declare one more instance of the JSP servlet, which has trimSpaces on, and to map that only for some pages, like those that produce binary, etc.

or the other way around, use trimSpaces as the default for the webapp and declare a jsp instance to be used without trimSpaces when needed.

so in the web.xml of your webapp you can do as below. also changing just in the webapp's web.xml does not require tomcat restart. maybe a clean of the webapp work dir, or at least a change of the jsp in question to force recompile.

i also found that Eclipse format source is not so good for JSP formatting as it is for Java. but more about that in another bug.

so for my part we can close this bug. i'll keep my trimSpaces on for a while for a whole webapp to see how it feels. but in any case, for isolated jsps that produce binaries, like comment 0 requests, this extra mapping of the jsp servlet solves the problem.

     <servlet>
        <servlet-name>jsp1</servlet-name>
        <servlet-class>org.apache.jasper.servlet.JspServlet</servlet-class>
        <init-param>
            <param-name>fork</param-name>
            <param-value>false</param-value>
        </init-param>
        <init-param>
            <param-name>xpoweredBy</param-name>
            <param-value>false</param-value>
        </init-param>
        <load-on-startup>3</load-on-startup>
                <init-param>
        <param-name>trimSpaces</param-name>
        <param-value>true</param-value>
    </init-param>
        
    </servlet>
    
       <servlet-mapping>
        <servlet-name>jsp1</servlet-name>
        <url-pattern>/index1.jsp</url-pattern>
    </servlet-mapping>

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 4 May 2010 14:03 UTC**
Just want to add, I revoke my initial claim that *Makumba* should provide this. I think this is something nice to have, but shouldn't be part of the Makumba taglib, as it has not much to do with the core strengths of Makumba.


Finally, taking the Eclipse JSP formatter as a kind-of reference to tailor our requirements for doesn't seem appropriate, as it imho does very inferior formatting anyhow when it comes to e.g. div or span tags.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 4 May 2010 21:10 UTC**
(In reply to comment #7)
> Just want to add, I revoke my initial claim that *Makumba* should provide this.
> I think this is something nice to have, but shouldn't be part of the Makumba
> taglib, as it has not much to do with the core strengths of Makumba.

great, so you have JSP trimSpaces, can we close this bug? 

if not, and if you want to do more search, note that HTML-level whitespace cleaners (which you will find a lot of) will not help.
 
> Finally, taking the Eclipse JSP formatter as a kind-of reference to tailor our
> requirements for doesn't seem appropriate, as it imho does very inferior
> formatting anyhow when it comes to e.g. div or span tags.

indeed it's weak. it was very good on the page I was testing with.

still, putting two successive mak:values on different lines is not unexpected from such a tool or indeed for manual formatting.

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 14:02 UTC**

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 20 Mar 2013 11:04 UTC**
this is a RFE outside makumba
